From a072217e86dfae3674c781347f57ac0ea9df55e3 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:49:46 -0400
Subject: [fs] nfsd: buffer-length check for SUPPATTR_EXCLCREAT

Message-id: <1403560314-32525-45-git-send-email-bfields@redhat.com>
Patchwork-id: 84555
O-Subject: [Patch RHEL7 044/172] nfsd4: buffer-length check for SUPPATTR_EXCLCREAT
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

This was an omission from 8c18f2052e756e7d5dea712fc6e7ed70c00e8a39
"nfsd41: SUPPATTR_EXCLCREAT attribute".

Cc: Benny Halevy <bhalevy@primarydata.com>
Cc: stable@vger.kernel.org
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: de3997a7eeb9ea286b15879fdf8a95aae065b4f7
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index 63f2395..668bfe1 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -2483,6 +2483,8 @@ out_acl:
    goto out;
  }
  if (bmval2 & FATTR4_WORD2_SUPPATTR_EXCLCREAT) {
+  if ((buflen -= 16) < 0)
+   goto out_resource;
   WRITE32(3);
   WRITE32(NFSD_SUPPATTR_EXCLCREAT_WORD0);
   WRITE32(NFSD_SUPPATTR_EXCLCREAT_WORD1);
-- 
1.7.1